test/managedwriter/writer_client_test.ts

```diff
@@ -0,0 +1,371 @@
+  })
+  describe('Common methods', () => {
+    it('should create a client without arguments', () => {
+      const client = new bigquerywriterModule.managedwriter.WriterClient();
+      client.setParent(projectId, datasetId, tableId);
+      assert(client);
+      assert.strictEqual(client.getParent(), parent);
+      assert(client.getClient());
+      assert.strictEqual(client.getWriteStreamType(), 'TYPE_UNSPECIFIED');
+    });
+
+    it('should create a client with arguments: parent, client, opts, writeStream', async () => {
+      const client = new bigquerywriterModule.managedwriter.WriterClient(
+        parent,
+        bqWriteClient,
+        clientOptions,
+        writeStreamType
+      );
+      assert(client);
+      assert.strictEqual(client.getParent(), parent);
+      assert(client.getClient());
+      const clientId = await client.getClient().getProjectId()
+      assert.strictEqual(clientId, clientOptions.projectId);
+      assert.strictEqual(client.getWriteStreamType(), 'TYPE_UNSPECIFIED');
+    });
+  });
+
+  describe('initializeStreamConnection', () => {
+    it('should invoke initalizeStreamConnection with or without clientOptions without errors', async () => {
+      await bqWriteClient.initialize();
+      writeStreamType = 'PENDING';
+      const client = new bigquerywriterModule.managedwriter.WriterClient(
+        parent,
+        bqWriteClient,
+        undefined,
+        writeStreamType
+      );
+      const numConnections: number =
+        client.getConnections().connection_list.length;
+
+      client.initializeStreamConnection().then(() => {
+        const streamId = 'fake-stream-id';
+        const streamIdResult = sinon.replace(
+          client,
+          'getStreamId',
+          sinon.fake.returns(streamId)
+        );
+        assert(client.getConnections().connection_list.length === 1);
+        assert(client.getConnections().connections['streamId']);
+        assert.strictEqual(streamIdResult, streamId);
+      });
+
+      /* invokes initializeStreamConnection with CallOptions */
+      const callOptions: gax.CallOptions = {};
+      const streamCallOptionsId = 'fake-stream-id-with-call-options';
+      client.initializeStreamConnection(callOptions).then(() => {
+        const streamIdCallOptionsResult = sinon.replace(
+          client,
+          'getStreamId',
+          sinon.fake.returns(streamCallOptionsId)
+        );
+        console.log(client.getConnections().connection_list.length);
+        assert(
+          client.getConnections().connection_list.length === numConnections + 2
+        );
+        assert(client.getConnections().connections['streamId']);
+        assert.strictEqual(streamIdCallOptionsResult, streamCallOptionsId);
+      });
+    });
+
+    /*it('should invoke initalizeStreamConnection with errors', () => {
+
+    })
+    it('should invoke initalizeStreamConnection with closed client', () => {
+
+    })*/
+  });
+
+  describe('appendRowsToStream', () => {
```

Comment:
We need two versions of this test, one for using an user created WriteStream, and another one using the DefaultStream. 

```
it('should invoke appendRowsToStream without errors', async () => {
     /* 
       Client and Proto initialization, rows creation, etc 
     ....
    */
     
      const streamName = await client.createWriteStream();
      const appendRowsResponsesResult: AppendRowsResponse[] = [
        {
          appendResult: {
            offset: offset,
          },
          writeStream: streamName,
        },
      ];
      try {
        const managedStream = await client.createManagedStream(
          streamName,
          protoDescriptor
        );
        const pw = await managedStream.appendRows(
          {
            serializedRows: [serializedRow1Message, serializedRow2Message],
          },
          offset
        );
        const result = await pw.getResult();
        const responses: AppendRowsResponse[] = [
          {
            appendResult: result.appendResult,
            writeStream: result.writeStream,
          },
        ];

        assert.deepEqual(appendRowsResponsesResult, responses);

        const rowCount = await managedStream.finalize();
        managedStream.close();
        assert.equal(rowCount, 2);

        const commitResponse = await client.batchCommitWriteStream({
          parent: client.getParent(),
          writeStreams: [streamName],
        });
        assert.equal(commitResponse.streamErrors?.length, 0);
      } finally {
        client.close();
      }

      return Promise.resolve();
    });
```    

For Default Stream

```
it('should invoke appendRows to default stream without errors', async () => {
        /* 
       Client and Proto initialization, rows creation, etc 
      ...
    */

      const appendRowsResponsesResult: AppendRowsResponse[] = [
        {
          appendResult: {
            offset: null,
          },
          writeStream: parent + '/streams/_default',
        },
      ];
      try {
        const managedStream = await client.createManagedStream(
          bigquerywriterModule.managedwriter.DefaultStream,
          protoDescriptor
        );
        const pw = await managedStream.appendRows({
          serializedRows: [serializedRow1Message, serializedRow2Message],
        });
        const result = await pw.getResult();
        const responses: AppendRowsResponse[] = [
          {
            appendResult: result.appendResult,
            writeStream: result.writeStream,
          },
        ];

        assert.deepEqual(appendRowsResponsesResult, responses);

        managedStream.close();
        client.close();
      } finally {
        client.close();
      }

      return Promise.resolve();
    });  
```

Reply:
We can also test appending rows multiple times and waiting for the proper response:

```
const appendRowsResponsesResult: AppendRowsResponse[] = [
        {
          appendResult: {
            offset: null,
          },
          writeStream: parent + '/streams/_default',
        },
        {
          appendResult: {
            offset: null,
          },
          writeStream: parent + '/streams/_default',
        },
      ];
      try {
        const managedStream = await client.createManagedStream(
          bigquerywriterModule.managedwriter.DefaultStream,
          protoDescriptor
        );
        const pw1 = await managedStream.appendRows({
          serializedRows: [serializedRow1Message, serializedRow2Message],
        });
        const pw2 = await managedStream.appendRows({
          serializedRows: [serializedRow1Message, serializedRow2Message],
        });
        const results = await Promise.all([
          pw1.getResult(),
          pw2.getResult(),
        ]);
        const responses: AppendRowsResponse[] = results.map(result => (
          {
            appendResult: result.appendResult,
            writeStream: result.writeStream,
          }
        ));

        assert.deepEqual(appendRowsResponsesResult, responses);

        managedStream.close();
        client.close();
      } finally {
        client.close();
      }

      return Promise.resolve();
 ```